### PR TITLE
Add New CloudService Resource in Azure SDK

### PR DESCRIPTION
**Pull request by ashvermamsft**
We already have a merged PR Azure/azure-rest-api-specs#11704 for swagger changes, and a follow up PR https://github.com/Azure/azure-rest-api-specs/pull/11950 to update the json files for the version package-2020-10-01-preview. We are starting this PR to merge the SDK changes.

itr1:

Generated Model and Operations files from swagger (using public repo)
Added tests for clousServices resource
Added session records and metadata files

### CODE REVIEW

**Comment by ashvermamsft** [comment]
@markcowl @grizzlytheodore please help with the review

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs:60]: Does this have all the properties of VirtualMachineExtension (i.e. is is a superset of properties)?

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs:57]: Is this a change in service behavior, or a correction to match service behavior in docs?

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:77]: Breaking change.  There needs to be an overload of this method that accepts VirtualMachineExtension

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:108]: Breaking change,a s above

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs:45]: Same comments as above

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:242]: Breaking change - need to include an overload that does not contain the ForceDeletion parameter to maintain backward compatibility

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1199]: Same comment

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1222]: same comment

**Review by markcowl — CHANGES_REQUESTED** [review]
Some breaking changes that need to be corrected, I think each can be fixed with simple customizations.

Also, you need to include the file generated by the generation script in /eng.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs:60]: some properties are different. But this file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020). By Customization Do you mean we can modify the generated files?

**Comment by ashvermamsft** [comment]
> Some breaking changes that need to be corrected, I think each can be fixed with simple customizations.
> 
> Also, you need to include the file generated by the generation script in /eng.

@markcowl  These files has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020). By Customization Do you mean we can modify the generated files? I haven't modified any generated files and have added only tests. Let me know if update in the generated file is allowed, will make the changes
About the metadat file from generation script, will include once the swagger PR is merged, and generation is done from there.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs:57]: This file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020).

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:77]: This file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020).

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:108]: This file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020).

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs:45]: This file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020).

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:242]: Again this file has been generated by autorest tool using the last published swagger json (i.e. 30-06-2020). If we are allowed to make changes to generated files, will make the changes. please confirm

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1199]: please confirm if generated files can be edited.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1222]: please confirm if generated files can be edited.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs
Comment (Sandido) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostGroup.cs:57]: This is possibly a bug in Autorest. I have seen this pop up in a number of Microsoft docs and public facing docs.

**Comment by ashvermamsft** [comment]
@markcowl  as @JeffreyRichter have approved the changes in this rest api pr (https://github.com/Azure/azure-rest-api-specs/pull/10745) , Can you please help in merging this PR? Please let me know if any additional steps are required to merge this PR.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:242]: Whether this is generated by AutoRest or not, we need to prevent breaking changes.  In this case, this is a breaking change in the SDK, which can be easily fixed with a single overload.  You should add this overload, or you will need breakign change approval.

Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:242]: To be clear on how to implement this:  This is an extension - you would define a new static class in the same namespace and simply add an extension method overload that matches the old pattern.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1222]: No, instead you add a separate file with a static class that implements the extension method overload with the older calling pattern.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs:341]: This is also a breakign change that can be mitigated with an additional partial class.  Here a is an example of how this can be mitigated: 

``` c# 
public partial class VirtualMachineOperations
{
  public Task<AzureOperationResponse> DeleteWithHttpMessagesAsync(string resourceGroupName, string vmName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
  {
    return DeleteWithHttpMessagesAsync( resourceGroupName, vmName, false, customHeaders, cancellationToken);
  }
```
}

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs:2831]: same comment

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperationsExtensions.cs:45]: This one is covered by the breaking change in the api, no further action is necessary.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:108]: This one is covered by the approved breaking change in the API, no further action is necessary

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMExtensionsOperations.cs:77]: This one is covered by the approved breaking change in the api.  No further action is necessary.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs
Comment (markcowl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetVMExtensionsOperations.cs:60]: This change is covered by the approved breaking change.

**Review by markcowl — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:242]: Added customizations as suggested.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs:341]: Added customizations as suggested.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs:2831]: Added customizations as suggested.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/IVirtualMachinesOperations.cs
Comment (markcowl) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/IVirtualMachinesOperations.cs:33]: So, this isn't quite right.  We have to be explicit about the existing overloads, rather than use default parameter values, so instead, the custom interface and implementation should have two new methods instead of one:

``` c#
Task<AzureOperationResponse> DeleteWithHttpMessagesAsync(string resourceGroupName, string vmName, Dictionary<string, List<string>> customHeaders);
Task<AzureOperationResponse> DeleteWithHttpMessagesAsync(string resourceGroupName, string vmName, Dictionary<string, List<string>> customHeaders, CancellationToken cancellationToken);
```

This pattern should be followed throughout.  Without this, there are a few combinations 
like ```DeleteWithHttpMessagesAsync("rg", "vm")```,  ```DeleteWithHttpMessagesAsync("rg", "vm", null)``` that cannot be disambiguated at runtime.

**Review by markcowl — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperationsExtensions.cs:1222]: Done

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/IVirtualMachinesOperations.cs
Comment (ashvermamsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/IVirtualMachinesOperations.cs]: Added as suggested.

**Comment by bquantump** [comment]
Please look into CI failures or we will close this.

**Comment by ashvermamsft** [comment]
@markcowl / @allenjzhang / !@bquantump In the latest iteration I have generated the full SDK from swagger repo master branch (using generate.ps1), haven't done any altering with the generated code, but still see the CI failure. Have included the metadata file as well. Any action/resolution needed from my end?

**Comment by markcowl** [comment]
@ashvermamsft According to the logs, there are changes in the following files when it generates from the master branch:
sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHost.cs
sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DedicatedHostUpdate.cs
sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/RollingUpgradePolicy.cs
sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/SdkInfo_ComputeManagementClient.cs

that are not reflected in this PR.  This could mean that either these files were changed but not included in the PR, that generation did not occur from the latest commit in master, or that there is an issue with the comparison.  If you can rule out (1) and (2), we cna look at (3)

**Comment by allenjzhang** [comment]
I did verification on the discrepancy, things look good. Merging.

**Comment by Sandido** [comment]
Includes the work for: 
https://github.com/Azure/azure-sdk-for-net/pull/17957/files
https://github.com/Azure/azure-rest-api-specs/pull/11202 
https://github.com/Azure/azure-rest-api-specs/pull/12521

**Pull request merged**
